src/Oracle/Oracle.cs

```diff
@@ -0,0 +1,358 @@
+            engine.LoadScript(NativeContract.Oracle.Script, CallFlags.None, 0);
+            engine.LoadScript(witnessDict[NativeContract.Oracle.Hash].InvocationScript, CallFlags.None);
+            if (engine.Execute() != VMState.HALT) return null;
+            tx.NetworkFee += engine.GasConsumed;
+
+            // Base size for transaction: includes const_header + signers + attributes + script + hashes
+
+            int size = Transaction.HeaderSize + tx.Signers.GetVarSize() + tx.Attributes.GetVarSize() + tx.Script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length);
+            size += witnessDict[NativeContract.Oracle.Hash].Size;
+            int size_inv = 66 * m;
+            size += IO.Helper.GetVarSize(size_inv) + size_inv + oracleSignContract.Script.GetVarSize();
+            var networkFee = ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * m;
+            using (ScriptBuilder sb = new ScriptBuilder())
+                networkFee += ApplicationEngine.OpCodePrices[(OpCode)sb.EmitPush(m).ToArray()[0]];
+            networkFee += ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * n;
+            using (ScriptBuilder sb = new ScriptBuilder())
+                networkFee += ApplicationEngine.OpCodePrices[(OpCode)sb.EmitPush(n).ToArray()[0]];
+            networkFee += ApplicationEngine.OpCodePrices[OpCode.PUSHNULL] + ApplicationEngine.ECDsaVerifyPrice * n;
+            tx.NetworkFee += networkFee;
+            tx.NetworkFee += size * NativeContract.Policy.GetFeePerByte(snapshot);
+
+            // Calcualte system fee
+
+            var request = NativeContract.Oracle.GetRequest(snapshot, response.Id);
+            tx.SystemFee = request.GasForResponse - tx.NetworkFee;
+
+            return tx;
+        }
+
+
+        public void AddResponseTxSign(StoreView snapshot, ulong requestId, byte[] sign, ECPoint oraclePub, Transaction respnoseTx = null)
+        {
+            var task = PendingQueue.GetOrAdd(requestId, new OracleTask
+            {
+                Id = requestId,
+                Request = NativeContract.Oracle.GetRequest(snapshot, requestId),
+                Signs = new SortedDictionary<ECPoint, byte[]>(),
+            });
+            task.Signs.TryAdd(oraclePub, sign);
+            if (respnoseTx != null)
+            {
+                task.Tx = respnoseTx;
+                var data = task.Tx.GetHashData();
+                task.Signs.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.Signs.Remove(p.Key));
+            }
+            else if (task.Tx != null)
+            {
+                var data = task.Tx.GetHashData();
+                if (!Crypto.VerifySignature(data, sign, oraclePub))
+                    throw new RpcException(-100, "Invalid response transaction sign");
+            }
+
+            ECPoint[] nodes = NativeContract.Oracle.GetOracleNodes(snapshot);
+            int m = nodes.Length - (nodes.Length - 1) / 3;
```

Comment:
This `M` is like the byzantine 2f+1, right?

Reply:
Maybe there should be some check on minimum nodes.Length here https://github.com/neo-project/neo/blob/56234cce47800b3e3ceb31b1a68d1580ec19ed6b/src/neo/SmartContract/Native/Oracle/OracleContract.cs#L171

Reply:
Done, Have added `CheckOracleAccount` method.